### DISCUSSION TITLE
x-args

### DISCUSSION THREAD

**Discussion by JeremyJaydan** [alpinejs/alpine#1999] [3. Feature ideas]
When using templates server-side, I'm often referencing variables within alpine methods that exist in a parent template.

I think it could be useful to specify defaults for `x-data` properties, similar to how function arguments can have defaults:
```javascript
function someFn(value = 'foobar'){
  return value;
}
```

The equivalent in Alpine could be a directive `x-args` or a modifier `x-data.args` where anything you specify works the same as `x-data` but can be overwritten by a parent `x-data`.

I think could make it even easier to build and share components too

**Comment by JeremyJaydan** [comment]
I wrote a directive a few days ago that handles this use case called `x-scope` that works similar to the above requirements however it's experimental & I've shelved it for now: <https://gist.github.com/JeremyJaydan/dd7f7df6830ec60acbcbcd9194c358be>

If you'd like to play with it, this is how it works:

If you specify `x-scope` on an element, it will "scope" the element so that it can't access parent data.
```html
<section
    x-data="{
       foobar: 'hello, world' 
    }">
    
    <div x-scope>
        <!-- won't work: -->
        <span x-text="foobar"></span>
    </div>
    
</section>
```

If you want to access parent data or simply want to specify defaults, it uses class syntax:
```html
<section
    x-data="{
       foobar: 'hello, world' 
    }">
    
    <div x-scope="{
        foobar = 'test';
    }">
        
        <!-- is 'hello, world': -->
        <span x-text="foobar"></span>
        
    </div>
    
</section>
```

It may or may not work completely, use at your own peril.

**Comment by SimoTod** [comment]
Alpine v3 can already access the parent scope without workaround. I might misunderstand the requirement. Are you trying to stop the default behaviour? are you trying to read values from parent without writing to the parent later? are you trying to give fallback values in the child in case the parent is not defined? A bit of both?
Probably a custom directive is the way to go since it seems a bit specific for your way to build pages

**Reply by JeremyJaydan** [reply]
> Are you trying to stop the default behaviour?

not my main goal, though I think stopping the default behaviour and only letting in explicitly defined variables within children would help with readability within server side templates

> are you trying to read values from parent without writing to the parent later?

not this, I want the parent variables to be the source of truth, child variables would just be fallbacks

> are you trying to give fallback values in the child in case the parent is not defined?

mostly this ^

> A bit of both?

yeah kinda, I like the idea of stopping parent variables if they're not explicitly defined in the children however I'm mainly thinking about fallbacks.

Overall it's really just for readability and redundancy for server side templates. A non-official directive could achieve this as I've commented separately above, just thought I'd share the use case.
